### PR TITLE
Adding support for upper and lower bound functions in SSA

### PR DESCRIPTION
**Pull request by Gamrix**
This diff adds support for `torch.nonzero` as a torch function that has both an upper and a lower bound size for shape functions, and therefore generates new symbolic dimensions in JIT.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77389
* #77388

### CODE REVIEW

**Comment by Gamrix** [comment]
@miladm Adding support for is_dynamic for torch.nonzero .

**Review by eellison — COMMENTED** [review]
Looking good! A few comments

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 1]
```diff
@@ -237,34 +242,58 @@ void transformShapeFunction(
   }
 }
 
-void registerSchema(
+std::shared_ptr<Graph> addFunctionToCache(
     const FunctionSchema* schema_string,
     const std::string& shape_compute_function_name,
     std::unordered_map<std::string, std::shared_ptr<Graph>>& reused_functions,
     const CompilationUnit& module) {
+  // TODO: When elias edits transformShapeFunction, remove the unnecessary
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:250]: Thought this diff would remove the need for the schema string, but clearly Ia m wrong.

https://github.com/pytorch/pytorch/pull/77120/files

File: torch/csrc/jit/passes/symbolic_shape_analysis.cpp [hunk 2]
```diff
@@ -747,6 +754,24 @@ std::shared_ptr<Graph> PropagateShapesWithShapeFunction(
   return op_analyzer.getShapeComputeGraph();
 }
 
+c10::SymbolicShape combine_bounds(
+    c10::SymbolicShape& lower_bound,
+    c10::SymbolicShape& upper_bound) {
+  TORCH_INTERNAL_ASSERT(lower_bound.rank() == upper_bound.rank());
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/passes/symbolic_shape_analysis.cpp:760]: I'll add a TODO for that. I don't think there are many cases in SSA where we will have different upper and lower number of dimmensions.

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 3]
```diff
@@ -237,34 +242,58 @@ void transformShapeFunction(
...
-  reused_functions[shape_compute_function_name] = graph;
+}
+
+void registerBoundedSchema(
+    const FunctionSchema* schema_string,
+    const std::string& lower_bound_function_name,
+    const std::string& upper_bound_function_name,
+    std::unordered_map<std::string, std::shared_ptr<Graph>>& reused_functions,
+    const CompilationUnit& module) {
+  auto lower_graph = addFunctionToCache(
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:292]: I renamed the function to reduce confusion.

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 4]
```diff
@@ -237,34 +242,58 @@ void transformShapeFunction(
...
-  transformShapeFunction(schema_string, graph);
-  // NB: we lint the shape functions registered in source
-  // in a test file
-  // LintShapeComputeGraph(schema_string, graph);
+void registerSchema(
+    const FunctionSchema* schema_string,
+    const std::string& shape_compute_function_name,
+    std::unordered_map<std::string, std::shared_ptr<Graph>>& reused_functions,
+    const CompilationUnit& module) {
+  auto graph = addFunctionToCache(
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:280]: I renamed the function to reduce confusion.

**Comment by Gamrix** [comment]
Updated to address comments

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**